### PR TITLE
Introduce the BlockClock control

### PR DESCRIPTION
**Pull request by shaavan**
- This PR introduces the Block clock control and adds it to the stub.qml to display.
- This PR implements the syncing state of the Block Clock.
- The implementation follows the [design guide](https://bitcoincore.app/block-clock/) and uses [this](https://stupefied-jones-dd209f.netlify.app/) prototype as the reference.

**Screenshot:**

Screenshot from 2022-08-09 17-43-03

Things to note for the current implementation:

Since this is the prototype for the Block clock, I focused on the implementation part and took some liberties with the nuances, which I would like to document here.

1. The color values for the: **Remaining time** text and **Background color of the block clock ring** didn't match the exact color values in the Theme.qml file. So I have used the closet matching Theme.color values to avoid hardcoded values and maintain consistency between light and dark mode.
2. Since there was no "yellow" color in the Theme.qml, I have used a hardcoded value for the progress indicator.
3. I have used the "bitcoin app" icon instead of the Black and White bitcoin icon, as this required additions to the resources and would be best done in a separate PR.
4. I have used static dots as the peer indicator. PR #127 introduces an implementation of the peer indicator and has a largely positive review. This PR could be updated with #127's implementation when that gets merged.

### CODE REVIEW

File: src/Makefile.qt.include [hunk 1]
```diff
@@ -312,6 +312,7 @@ QML_RES_QML = \
   qml/components/StorageLocations.qml \
   qml/components/StorageOptions.qml \
   qml/components/StorageSettings.qml \
+  qml/controls/BlockClock.qml \
```
Comment (jarolrod) [inline src/Makefile.qt.include:315]: this is a component. not a control

File: src/qml/controls/BlockClock.qml [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+       id: animationArcEnd
+       enabled: true
+       NumberAnimation {
+            easing.type: Easing.Bezier
+            easing.bezierCurve: [0.5, 0.0, 0.2, 1, 1, 1]
+            duration: root.animationDuration
+        }
+    }
+
+    Canvas {
```
Comment (jarolrod) [inline src/qml/controls/BlockClock.qml:38]: instead of canvas, this should be the simpler ShapePath: https://doc.qt.io/qt-6/qml-qtquick-shapes-shapepath.html

File: src/Makefile.qt.include (hunk 1 above)
Comment (shaavan) [reply src/Makefile.qt.include:315]: The next update of my PR intends to make the `BlockClock.qml` file a purely abstract structure and introduce a component file that uses it to paint different states of the block clock as needed. So I think moving `BlockClock.qml` to components now would be a wasted effort.

Should you still think I shall move the `BlockClock.qml` to controls, or if there is a better approach for achieving the final BlockClock state, I would love to hear that.

File: src/qml/controls/BlockClock.qml (hunk 2 above)
Comment (shaavan) [reply src/qml/controls/BlockClock.qml:38]: I looked through using the ShapePath to paint the BlockClock. However, using this would require us to introduce a new dependency, `qml-module-qtquick-shapes` for building the GUI.

So I think this suggestion needs further discussion before implementing it.

A friendly ping, @hebasto

What do you think about this suggestion?

Comment (promag) [reply src/qml/controls/BlockClock.qml:38]: Unfortunately, from my experiments, `ShapePath` and related items have issues. The most annoying issue happens with small curves where tessellation yields bad curve smoothness. Another issue is that with complex drawings you end up with lots of items and bindings everywhere, which is also not that performant.

Going forward a stack/composition of Canvas is the best option, like photoshop layers. For instance, the Canvas on the bottom could be static or updated less often than the one on the top.

File: src/qml/nodemodel.cpp [hunk 3]
```diff
@@ -44,6 +120,13 @@ void NodeModel::initializeResult([[maybe_unused]] bool success, interfaces::Bloc
     // TODO: Handle the `success` parameter,
     setBlockTipHeight(tip_info.block_height);
     setVerificationProgress(tip_info.verification_progress);
+
+    const CBlockIndex* tip;
+    node::NodeContext* context = m_node.context();
+    ChainstateManager* chainman = context->chainman.get();
```
Comment (jarolrod) [inline src/qml/nodemodel.cpp:126]: this is too low level of an object to be handling here when we have `interfaces::chain`

File: src/qml/nodemodel.h [hunk 4]
```diff
@@ -25,15 +29,24 @@ class NodeModel : public QObject
...
 public:
     explicit NodeModel(interfaces::Node& node);
 
     int blockTipHeight() const { return m_block_tip_height; }
     void setBlockTipHeight(int new_height);
+    int remainingSyncTime() const { return m_remaining_time; }
     double verificationProgress() const { return m_verification_progress; }
     void setVerificationProgress(double new_progress);
+    QVariantList blockTimeList() const { return m_block_time_list; }
+    bool setBlockTimeList(int new_block_time);
```
Comment (jarolrod) [inline src/qml/nodemodel.h:46]: why should the nodemodel, a proxy class for our internal node, keep something like a blocklist?

File: src/qml/controls/BlockClock.qml [hunk 5]
```diff
@@ -0,0 +1,160 @@
+import QtQuick 2.15
+import QtQuick.Controls 2.15
+import QtQuick.Layouts 1.15
+
+Item {
```
Comment (jarolrod) [inline src/qml/controls/BlockClock.qml:5]: This is the implementation of the block clock, and the file residing at `src/qml/components/BlockClockComponent.qml` is just a controller for this. A better architecture is to have a controller that loads the different states the block clock can be in.

You'd be able to split up this file, as it is very long, between a simple circular progress bar that shows the verification progress while in IBD, then another dial like progress bar for the blocks. Then load these into the controller.

File: src/qml/controls/BlockClock.qml [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    ColumnLayout {
+        anchors.centerIn: root
+        Image {
+            Layout.alignment: Qt.AlignCenter
+            source: "image://images/app"
+            sourceSize.width: 30
+            sourceSize.height: 30
+            Layout.bottomMargin: 15
+        }
+        Text {
```
Comment (jarolrod) [inline src/qml/controls/BlockClock.qml:130]: isn't this recreating our header component?

File: src/qml/nodemodel.cpp [hunk 7]
```diff
@@ -28,12 +29,87 @@ void NodeModel::setBlockTipHeight(int new_height)
...
+    }
+
+    double ratio_of_12_hour_passed = (new_block_time - time_at_12th_hour) / (double)sec_in_12_hours;
+
+    m_block_time_list.push_back(ratio_of_12_hour_passed);
+    Q_EMIT blockTimeListChanged();
+    return true;
+}
+
+void NodeModel::setBlockTimeListInitial(const CBlockIndex* pblockindex)
```
Comment (jarolrod) [inline src/qml/nodemodel.cpp:88]: Yeah this is definitely too low level for us to be passing around here

**Review by jarolrod — COMMENTED** [review]
Branch doesn't build

**Comment by shaavan** [comment]
Updated from 621eae1074365116d5bfd5e58eff2696eaf6a846 to 1f134fdb3bcc29213b55658a69d1d8bf72079554 ([pr148.02](https://github.com/shaavan/gui-qml/commits/pr148.02) -> [pr148.03](https://github.com/shaavan/gui-qml/commits/pr148.03), [diff](https://github.com/shaavan/gui-qml/compare/pr148.02..pr148.03))
Addressed @jarolrod comment

Update:
- Made changes to fix the problem of the branch not building. Thanks, @jarolrod, for bringing that to notice and helping figure out the root of the problem.

File: src/qml/nodemodel.cpp (hunk 3 above)
Comment (shaavan) [reply src/qml/nodemodel.cpp:126]: Let me take a look into replacing this low-level object handling with interfaces::chain

File: src/qml/nodemodel.h (hunk 4 above)
Comment (shaavan) [reply src/qml/nodemodel.h:46]: Considering your previous suggestion about using the `interfaces::chain`, I think I can create a separate .cpp file handling chain-related parameters such as blockTimeList

File: src/qml/controls/BlockClock.qml (hunk 5 above)
Comment (shaavan) [reply src/qml/controls/BlockClock.qml:5]: Let me try rephrasing your suggestion to ensure I understood it correctly.
You suggest I create a separate `component` file for each block clock state, create a controller file in `qml/control`, and switch to loading different block clock component files based on the requirement.

Have I understood it correctly?

File: src/qml/nodemodel.cpp (hunk 7 above)
Comment (shaavan) [reply src/qml/nodemodel.cpp:88]: Let me experiment with `interfaces::chain` and devise an approach that avoids low-level object handling.

**Comment by shaavan** [comment]
Updated from 23f8a833b57bbdad3177ff23751ec0cdfb8d5b09 to 621eae1074365116d5bfd5e58eff2696eaf6a846 ([pr148.03](https://github.com/shaavan/gui-qml/commits/pr148.03) -> [pr148.04](https://github.com/shaavan/gui-qml/commits/pr148.04))

Changes:
- Introduced a new chainmodel class to use the interfaces::chain and directly use the chain info without going too low level in the code.
- The new Block Clock displays the timezone-dependent block time data.
- Added property to Header.qml so it can be used directly in the BlockClock.qml file.

Here is the demo of the latest iteration of the block clock:

https://user-images.githubusercontent.com/85434418/204341167-fe8cb5a4-7189-4cbd-b2f8-1db93be3cf3d.mov

**Comment by shaavan** [comment]
Updated from 621eae1074365116d5bfd5e58eff2696eaf6a846 to 2a5e9983bebab10ce7be79eebdaf6104a992ca32 ([pr148.04](https://github.com/shaavan/gui-qml/commits/pr148.04) -> [pr148.05](https://github.com/shaavan/gui-qml/commits/pr148.05))

Changes: 
1. Rebased over main.
2. Added the **Connecting** and **Paused** state for the block clock.
3. Integrated the first three commits of #127. This allows the PeerIndicator dots to be animated and correctly show the peer connection state.


**Demo:**

https://user-images.githubusercontent.com/85434418/207072355-c7371de4-fdd5-4dbe-bfee-db74e9a61818.mov

**Comment by GBKS** [comment]
Thanks for the updates. It looks like the artifacts are not building due a lint issue (the Android build is only 22 bytes). Could you please address this? I'd love to get hands-on with this update.

File: .vscode/settings.json [hunk 8]
```diff
@@ -0,0 +1,7 @@
+{
+    "files.associations": {
+        "__split_buffer": "cpp",
+        "deque": "cpp",
+        "list": "cpp"
+    }
+}
```
Comment (hebasto) [inline .vscode/settings.json:7]: Why is this file required?

**Review by hebasto — COMMENTED** [review]
To fix the lint task, suggesting the patch as follows:
```diff
diff --git a/src/qml/chainmodel.h b/src/qml/chainmodel.h
index 9ab25f35c..2943c9c16 100644
--- a/src/qml/chainmodel.h
+++ b/src/qml/chainmodel.h
@@ -2,8 +2,8 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
-#ifndef BITCOIN_QML_CHAIN_MODEL_H
-#define BITCOIN_QML_CHAIN_MODEL_H
+#ifndef BITCOIN_QML_CHAINMODEL_H
+#define BITCOIN_QML_CHAINMODEL_H
 
 #include <interfaces/chain.h>
 
@@ -47,4 +47,4 @@ private:
     interfaces::Chain& m_chain;
 };
 
-#endif // BITCOIN_QML_CHAIN_MODEL_H
\ No newline at end of file
+#endif // BITCOIN_QML_CHAINMODEL_H
```

File: src/qml/controls/BlockClock.qml [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+                ctx.stroke()
+            }
+        }
+    }
+
+    ColumnLayout {
+        anchors.centerIn: root
+        Image {
+            Layout.alignment: Qt.AlignCenter
+            source: "image://images/app"
```
Comment (jarolrod) [inline src/qml/controls/BlockClock.qml:129]: This should be using the bitcoin circle icon introduced in https://github.com/bitcoin-core/gui-qml/pull/176. That's what is specified in the [design files](https://www.figma.com/file/GaCoOSNHB2yMB9ThiDtred/Bitcoin-Core-App?node-id=3259%3A73676&t=pjV9bY1q9KckaP4Q-0)

**Comment by shaavan** [comment]
Updated from 2a5e9983bebab10ce7be79eebdaf6104a992ca32 to 0f5461eef14e3ac0739d9b3e3448dd78c94ebff7 ([pr148.05](https://github.com/shaavan/gui-qml/commits/pr148.05) -> [pr148.06](https://github.com/shaavan/gui-qml/commits/pr148.06))

Addressed @hebasto @jarolrod @GBKS comments.

Changes:

1. Switch the Bitcoin app icon for the Bitcoin-circle icon.
2. Fixed `QObject::moveToThread: Cannot move objects with a parent` warning.
3. Correction Font size and margins according to [design docs](https://www.figma.com/file/ek8w3n3upbluw5UL2lGhRx/Bitcoin-Core-App-Design?node-id=616%3A0&t=poh7bpeInsyETLvD-0).
4. Removed .vscode/settings.json file.
5. Applied the patch suggested by @hebasto to fix the lint warning.

**Screenshots:**

|pr148.05|pr148.06|
|---------|---------|
|Screenshot 2022-12-15 at 11 51 04 PM|Screenshot 2022-12-15 at 11 48 44 PM|
|Screenshot 2022-12-15 at 11 52 02 PM|Screenshot 2022-12-15 at 11 49 19 PM|
|Screenshot 2022-12-15 at 11 51 32 PM|Screenshot 2022-12-15 at 11 49 51 PM|
|Screenshot 2022-12-15 at 11 51 30 PM|Screenshot 2022-12-15 at 11 49 43 PM|

File: .vscode/settings.json (hunk 8 above)
Comment (shaavan) [reply .vscode/settings.json]: The file has been removed in 0f5461e

File: src/qml/chainmodel.cpp [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+ChainModel::ChainModel(interfaces::Chain& chain)
+    : m_chain{chain}
+{
+    timer = new QTimer(this);
+    connect(timer, &QTimer::timeout, this, &ChainModel::setCurrentTimeRatio);
+    timer->start(1000);
+
+    QThread* timer_thread = new QThread;
+    timer->moveToThread(timer_thread);
```
[the remark this answers is not in the document]
Comment (shaavan) [reply src/qml/chainmodel.cpp:20]: Fixed in 0f5461e

File: src/qml/controls/BlockClock.qml (hunk 9 above)
Comment (shaavan) [reply src/qml/controls/BlockClock.qml]: Addressed in 0f5461e.

**Comment by shaavan** [comment]
Updated from 0f5461eef14e3ac0739d9b3e3448dd78c94ebff7 to 0fbe6c809eadabde0ca4c6f443dea60feae3fb34 ([pr148.06](https://github.com/shaavan/gui-qml/commits/pr148.06) -> [pr148.07](https://github.com/shaavan/gui-qml/commits/pr148.07))

- Rebased the branch over the main.

**Comment by shaavan** [comment]
Updated from [0fbe6c8 ](https://github.com/bitcoin-core/gui-qml/commit/0fbe6c809eadabde0ca4c6f443dea60feae3fb34)to [29f019e](https://github.com/bitcoin-core/gui-qml/commit/29f019e4a5278a4e7b5aaf56a49bee309bcc4d2a)

- Fixed the trailing whitespace lint failure.

**Comment by shaavan** [comment]
Updated from 29f019e4a5278a4e7b5aaf56a49bee309bcc4d2a to 8173bf6838997dce6dc6e908a54801fdcb5e0ad8 ([pr148.08](https://github.com/shaavan/gui-qml/commits/pr148.08) -> [pr148.09](https://github.com/shaavan/gui-qml/commits/pr148.09))

- Moved the simplified block_clock changes from #212 to this PR to preserve the comment history for the block_clock and prevent splitting of discussion in two separate PRs.

Differences as explained in #212 Description:

> 1. The PeersIndicator has been replaced with plain rectangles to keep the UI intact.
> 2. The conns property representing if the number of connections is greater than 0, has been set to always positive. With this > change, the "connecting" page, though implemented, becomes inaccessible through the UI.

File: src/qml/components/BlockClockComponent.qml [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+        State {
+            name: "blockClock"; when: synced && !pause && conns
+            PropertyChanges {
+                target: blockClock
+
+                ringProgress: blockList[0]
+                header: Math.floor(nodeModel.blockTipHeight / 1000) + "," + nodeModel.blockTipHeight % 1000
+                subText: "Blocktime"
+                blockList: chainModel.timeRatioList
```
Comment (johnny9) [inline src/qml/components/BlockClockComponent.qml:34]: QML error appears when shutting down.

```
2022-12-17T03:52:37Z Shutdown: In progress...
2022-12-17T03:52:37Z torcontrol thread exit
2022-12-17T03:52:37Z addcon thread exit
2022-12-17T03:52:37Z msghand thread exit
2022-12-17T03:52:37Z net thread exit
2022-12-17T03:52:39Z opencon thread exit
2022-12-17T03:52:39Z DumpAnchors: Flush 0 outbound block-relay-only peer addresses to anchors.dat started
2022-12-17T03:52:39Z DumpAnchors: Flush 0 outbound block-relay-only peer addresses to anchors.dat completed (0.01s)
2022-12-17T03:52:39Z scheduler thread exit
2022-12-17T03:52:39Z Writing 0 unbroadcast transactions to disk.
2022-12-17T03:52:39Z Dumped mempool: 3.7e-06s to copy, 0.00448635s to dump
2022-12-17T03:52:39Z Shutdown: done
2022-12-17T03:52:39Z GUI: qrc:/qml/components/BlockClockComponent.qml:34: TypeError: Cannot read property 'timeRatioList' of null
```

File: src/qml/controls/BlockClock.qml [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+                ctx.arc(x, y, (width / 2) - parent.lineWidth / 2, start, end, false)
+                ctx.lineWidth = root.lineWidth
+                ctx.strokeStyle = root.colorCircle
+                ctx.stroke()
+            }
+
+            else {
+                var del = 0.0025
+                // Paint Block time points
+                for (var i=1; i<parent.blockList.length - 1; i++) {
```
Comment (johnny9) [inline src/qml/controls/BlockClock.qml:96]: The following error appears when going into the paused state.

```
2022-12-17T03:50:17Z GUI: qrc:/qml/controls/BlockClock.qml:96: TypeError: Value is undefined and could not be converted to an object
```

Comment (johnny9) [reply src/qml/controls/BlockClock.qml:96]: Small style comment, operators should have space before/after for easier readability.

File: src/qml/pages/node/NodeRunner.qml [hunk 13]
```diff
@@ -18,26 +18,9 @@ Page {
...
-            ProgressIndicator {
-                width: 200
-                Layout.alignment: Qt.AlignCenter
-                progress: nodeModel.verificationProgress
-            }
-        }
+
+        Component.onCompleted: nodeModel.startNodeInitializionThread();
+
+        BlockClockComponent{}
```
Comment (johnny9) [inline src/qml/pages/node/NodeRunner.qml:24]:
```
GUI: qrc:/qml/pages/node/NodeRunner.qml:24:9: QML BlockClockComponent: Detected anchors on an item that is managed by a layout. This is undefined behavior; use Layout.alignment instead.
```
Should be able to remove the ColumnLayout containing the BlockClockComponent and just define the BlockClockComponent as the content to this Page.  When you do this, also remove the `anchors.fill: parent` and `spacing: 0` in this file.

File: src/qml/chainmodel.h [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+private:
+    QVariantList m_time_ratio_list{0.0};
+
+    QTimer* timer;
+
+    interfaces::Chain& m_chain;
+};
+
+#endif // BITCOIN_QML_CHAINMODEL_H
```
Comment (johnny9) [inline src/qml/chainmodel.h:50]: "No newline at end of file"

File: src/qml/chainmodel.cpp [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+void ChainModel::setCurrentTimeRatio()
+{
+    int secsSinceMeridian = (QTime::currentTime().msecsSinceStartOfDay() / 1000) % SECS_IN_12_HOURS;
+    double currentTimeRatio = double(secsSinceMeridian) / SECS_IN_12_HOURS;
+
+    if (currentTimeRatio < m_time_ratio_list[0].toDouble()) { // That means time has crossed a meridian
+        m_time_ratio_list.erase(m_time_ratio_list.begin() + 1, m_time_ratio_list.end());
+    }
+    m_time_ratio_list[0] = currentTimeRatio;
+}
```
Comment (johnny9) [inline src/qml/chainmodel.cpp:75]: "No newline at end of file"

**Review by johnny9 — CHANGES_REQUESTED** [review]
Views look nice and render correctly.  Tested pause/unpause as well as a full sync on signet at got the following results.

It looks great.  Tested pause/resume as well as a sync on signet and here's what I got.

|Screenshot_2022-12-16_22-18-31|Screenshot_2022-12-16_22-19-02|Screenshot_2022-12-16_22-25-01|Screenshot_2022-12-16_22-26-42|Screenshot_2022-12-16_22-28-51|
|---|---|---|---|---|

A few QML errors when running that need to be resolved before merging. I've left comments for those errors at the corresponding line numbers.

`tail -f debug.log | grep GUI` to watch for issues when running bitcoin-qt.

**Comment by shaavan** [comment]
Updated from 8173bf6838997dce6dc6e908a54801fdcb5e0ad8 to bcbc80b6666e43aa07ee46e09289b4a34f3fa30b ([pr148.09](https://github.com/shaavan/gui-qml/commits/pr148.09) -> [pr148.10](https://github.com/shaavan/gui-qml/commits/pr148.10))

Addressed @johnny9 comments.

Changes:

1. Fixed Type Errors as pointed out in [this](https://github.com/bitcoin-core/gui-qml/pull/148#pullrequestreview-1222038049) comment.
2. Added space before and after operator in BlockClock.qml
3. Added a new line at the end of the file in chainmodel.cpp and chainmodel.h.
4. Removed the redundant ColumnLayout in Noderunner.qml.

**Comment by hebasto** [comment]
The CI lint task error:
```
fatal: Invalid revision range fd4e57f63654e2d8e7d0de0aafc335715a53a5cc..c4e1ce262ffb5e6956a7b912f2dcf7560823682e
```

looks a bit weird as I failed to find a branch the c4e1ce262ffb5e6956a7b912f2dcf7560823682e commit belongs to.

I hope rebasing will help.

@shaavan And please make sure your local branch does not contain any merge commits.

**Comment by shaavan** [comment]
Updated from bcbc80b6666e43aa07ee46e09289b4a34f3fa30b to 68fa33b202e528ff0eae839f7163be8a7f939392  ([pr148.10](https://github.com/shaavan/gui-qml/commits/pr148.10) -> [pr148.11](https://github.com/shaavan/gui-qml/commits/pr148.11))

- Rebased PR over current main.

File: src/qml/controls/BlockClock.qml [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+                ctx.beginPath();
+                ctx.arc(x, y, (width / 2) - parent.lineWidth / 2, start, end, false)
+                ctx.lineWidth = root.lineWidth
+                ctx.strokeStyle = root.colorCircle
+                ctx.stroke()
+            }
+
+            else {
+                var del = 0.0025
+                if (parent.blockList.length) {
```
Comment (johnny9) [inline src/qml/controls/BlockClock.qml:95]: Still getting `2022-12-19T23:10:24Z GUI: qrc:/qml/controls/BlockClock.qml:95: TypeError: Value is undefined and could not be converted to an object` when going to the paused state. i think doing `if (parent.blockList) {` should resolve it.

**Review by johnny9 — CHANGES_REQUESTED** [review]
QML error when going into the paused state still exists. Other issues are resolved.

**Comment by shaavan** [comment]
Updated from 68fa33b202e528ff0eae839f7163be8a7f939392 to 20b4ed55f1b9417fa0a1457736be1b6c1c54f1dd ([pr148.11](https://github.com/shaavan/gui-qml/commits/pr148.11) -> [pr148.12](https://github.com/shaavan/gui-qml/commits/pr148.12), [diff](https://github.com/shaavan/gui-qml/compare/pr148.11..pr148.12))

Addressed @johnny9 comment.

- Fixed `TypeError: Value is undefined and could not be converted to an object` error.

- Also addressed one formatting issue, present in displaying Block Counter.

**Explaination:**

For example, for the block count of **122010**, difference:

|pr148.11|pr148.12|
|--------|---------|
|Screenshot 2022-12-20 at 5 54 00 PM|Screenshot 2022-12-20 at 6 16 46 PM|

**Comment by shaavan** [comment]
Updated from 20b4ed55f1b9417fa0a1457736be1b6c1c54f1dd to abb0bef2460b4ba98cf02758dedd052efe4ec722 ([pr148.12](https://github.com/shaavan/gui-qml/commits/pr148.12) -> [pr148.13](https://github.com/shaavan/gui-qml/commits/pr148.13), [diff](https://github.com/shaavan/gui-qml/compare/pr148.12..pr148.13))

- Modified `timeRatioList` so that blockClock start displaying from the "0" point.

**Screenshot:**

|pr148.12|pr148.13|
|--------|---------|
|Screenshot 2022-12-20 at 6 28 41 PM|Screenshot 2022-12-20 at 6 27 57 PM|

File: src/qml/pages/node/NodeRunner.qml (hunk 13 above)
Comment (shaavan) [reply src/qml/pages/node/NodeRunner.qml]: Addressed in bcbc80b6666e43aa07ee46e09289b4a34f3fa30b.

File: src/qml/chainmodel.cpp (hunk 15 above)
Comment (shaavan) [reply src/qml/chainmodel.cpp]: Addressed in bcbc80b6666e43aa07ee46e09289b4a34f3fa30b

File: src/qml/chainmodel.h (hunk 14 above)
Comment (shaavan) [reply src/qml/chainmodel.h]: Addressed in bcbc80b6666e43aa07ee46e09289b4a34f3fa30b

File: src/qml/components/BlockClockComponent.qml (hunk 11 above)
Comment (shaavan) [reply src/qml/components/BlockClockComponent.qml:34]: Addressed in bcbc80b6666e43aa07ee46e09289b4a34f3fa30b

File: src/qml/controls/BlockClock.qml (hunk 12 above)
Comment (shaavan) [reply src/qml/controls/BlockClock.qml]: Addressed in 20b4ed55f1b9417fa0a1457736be1b6c1c54f1dd

File: src/qml/controls/BlockClock.qml (hunk 16 above)
Comment (shaavan) [reply src/qml/controls/BlockClock.qml]: Addressed in 20b4ed55f1b9417fa0a1457736be1b6c1c54f1dd

Thanks for all of your reviews!

**Review by johnny9 — COMMENTED** [review]
Good work on this Shashwat. All of the QML errors appear to be resolved now. Views render as expected

File: src/node/interfaces.cpp [hunk 17]
```diff
@@ -514,6 +514,11 @@ class ChainImpl : public Chain
         LOCK(::cs_main);
         return Assert(chainman().ActiveChain()[height])->GetBlockHash();
     }
+    int64_t getBlockTime(int height) override
+    {
+        LOCK(::cs_main);
+        return Assert(chainman().ActiveChain()[height])->GetBlockTime();
```
Comment (hebasto) [inline src/node/interfaces.cpp:520]: fec6a0e9c7c4d98a88628ab57ce33144923da85d

In `Chain` interface, probably better to use a type-safe `CBlockIndex::Time()` https://github.com/bitcoin-core/gui-qml/blob/fd4e57f63654e2d8e7d0de0aafc335715a53a5cc/src/chain.h#L279-L282

Anyway, it will take a broader discussion when it will about to merge into the main repo.

File: src/qml/chainmodel.h [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    void setTimeRatioList(int new_time);
+    void setTimeRatioListInitial();
+
+Q_SIGNALS:
+    void timeRatioListChanged();
+
+private:
+    QVariantList m_time_ratio_list{0.0};
+
+    QTimer* timer;
```
Comment (hebasto) [inline src/qml/chainmodel.h:45]: a287f9e44660b80349defa8f34f14c8b9f0d8b78
Naming convention:
```suggestion
    QTimer* m_timer;
```

File: src/qml/chainmodel.cpp [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+ChainModel::ChainModel(interfaces::Chain& chain)
+    : m_chain{chain}
+{
+    timer = new QTimer();
+    connect(timer, &QTimer::timeout, this, &ChainModel::setCurrentTimeRatio);
+    timer->start(1000);
+
+    QThread* timer_thread = new QThread;
+    timer->moveToThread(timer_thread);
+    timer_thread->start();
```
[the remark this answers is not in the document]
Comment (shaavan) [reply src/qml/chainmodel.cpp:21]: The **currentTimeRatio** represents the ratio of 12 hours that has been passed.
So, for example, at 4 am or pm, the currenTimeRatio would be 0.3.

We need to update this information in the blockClock each second so that the blockClock always shows the correct blockTime.

So that's why I created a separate 1-sec timer thread to call this function each second and update the **currentTimeRatio**.

File: src/node/interfaces.cpp (hunk 17 above)
Comment (shaavan) [reply src/node/interfaces.cpp:520]: Sure, let me look through it.

> it will take a broader discussion when it will about to merge into the main repo.

Let me include that, along with other future suggestions for this file.

**Comment by shaavan** [comment]
Updated from abb0bef2460b4ba98cf02758dedd052efe4ec722 to 67d12a67536b636bbda3f5bb8d1e62cc5fa66d65 ([pr148.13](https://github.com/shaavan/gui-qml/commits/pr148.13) -> [pr148.14](https://github.com/shaavan/gui-qml/commits/pr148.14), [diff](https://github.com/shaavan/gui-qml/compare/pr148.13..pr148.14))

Addressed @hebasto comments.

Changes:

- Changed 1-sec QTimer from a global variable to a local variable.
- Changed names of variables mentioned in [this](https://github.com/bitcoin-core/gui-qml/pull/148#discussion_r1053596002) comment from camel-style naming to snake-style naming. For full diff, check [diff](https://github.com/shaavan/gui-qml/compare/pr148.13..pr148.14).
- Added descriptive comment for `m_time_ratio_list`, and the values it stores.

File: src/qml/chainmodel.cpp [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+
+    m_time_ratio_list.push_back(double(new_time - timeAtMeridian) / SECS_IN_12_HOURS);
+
+    Q_EMIT timeRatioListChanged();
+}
+
+int ChainModel::timestampAtMeridian()
+{
+    int secsSinceMeridian = (QTime::currentTime().msecsSinceStartOfDay() / 1000) % SECS_IN_12_HOURS;
+    int currentTimestamp = QDateTime::currentSecsSinceEpoch();
```
[the remark this answers is not in the document]
Comment (shaavan) [reply src/qml/chainmodel.cpp]: Addressed in 67d12a67536b636bbda3f5bb8d1e62cc5fa66d65

File: src/qml/chainmodel.cpp [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+
+    return currentTimestamp - secsSinceMeridian;
+}
+
+void ChainModel::setTimeRatioListInitial()
+{
+    int timeAtMeridian = timestampAtMeridian();
```
[the remark this answers is not in the document]
Comment (shaavan) [reply src/qml/chainmodel.cpp]: Addressed in 67d12a67536b636bbda3f5bb8d1e62cc5fa66d65

File: src/qml/chainmodel.cpp [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+        m_time_ratio_list.push_back(double(m_chain.getBlockTime(height) - timeAtMeridian) / SECS_IN_12_HOURS);
+    }
+
+    Q_EMIT timeRatioListChanged();
+}
+
+void ChainModel::setCurrentTimeRatio()
+{
+    int secsSinceMeridian = (QTime::currentTime().msecsSinceStartOfDay() / 1000) % SECS_IN_12_HOURS;
+    double currentTimeRatio = double(secsSinceMeridian) / SECS_IN_12_HOURS;
```
[the remark this answers is not in the document]
Comment (shaavan) [reply src/qml/chainmodel.cpp]: Addressed in 67d12a67536b636bbda3f5bb8d1e62cc5fa66d65

File: src/qml/chainmodel.h [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+
+public Q_SLOTS:
+    void setTimeRatioList(int new_time);
+    void setTimeRatioListInitial();
+
+Q_SIGNALS:
+    void timeRatioListChanged();
+
+private:
+    QVariantList m_time_ratio_list{0.0};
```
[the remark this answers is not in the document]
Comment (shaavan) [reply src/qml/chainmodel.h:49]: Addressed in 67d12a67536b636bbda3f5bb8d1e62cc5fa66d65

File: src/qml/chainmodel.h (hunk 18 above)
Comment (shaavan) [reply src/qml/chainmodel.h]: Addressed in 67d12a67536b636bbda3f5bb8d1e62cc5fa66d65

Since this timer didn't need to be a global variable, I changed it to a local variable.

**Comment by hebasto** [comment]
Closing in favor of #220.
